clippy_lints/src/unused_unit.rs

```diff
@@ -109,6 +109,7 @@ impl<'tcx> LateLintPass<'tcx> for UnusedUnit {
             && let AssocItemConstraintKind::Equality { term: Term::Ty(hir_ty) } = constraints[0].kind
             && args.span_ext.hi() != poly.span.hi()
             && !hir_ty.span.from_expansion()
+            && args.span_ext.hi() != hir_ty.span.hi()
```

Comment:
Perhaps this should have been covered by `!hir_ty.span.from_expansion()`. Certainly, there are no explicit `-> ()` tokens in the source. If this is the case, I'm happy to close this PR and file an issue upstream.

Reply:
Those expansion checks are quite finicky; sometimes you have an expression without expansion but a subexpression expanded, etc. So as long as this appears to fix the issue, it's better than nothing.